hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFs.java

```diff
@@ -915,11 +915,24 @@ public FileStatus getFileLinkStatus(final Path f)
       if (inode.isLink()) {
         INodeLink<AbstractFileSystem> inodelink = 
           (INodeLink<AbstractFileSystem>) inode;
-        result = new FileStatus(0, false, 0, 0, creationTime, creationTime,
+        Path linkedPath = new Path(inodelink.targetDirLinkList[0].toString());
+        ChRootedFs linkedFs = (ChRootedFs) inodelink.getTargetFileSystem();
+        try {
+          FileStatus status = linkedFs.getMyFs().getFileStatus(linkedPath);
+          result = new FileStatus(status.getLen(), false,
+            status.getReplication(), status.getBlockSize(),
+            status.getModificationTime(), status.getAccessTime(),
+            status.getPermission(), status.getOwner(), status.getGroup(),
+            inodelink.getTargetLink(),
+            new Path(inode.fullPath).makeQualified(
+                myUri, null));
+        } catch (FileNotFoundException ex) {
```

Comment:
similar comment regarding FileNotFoundException. I think in general, it's better to match behavior of non-federated client. If a path does not exist, just throw back FileNotFoundException.

Reply:
> similar comment regarding FileNotFoundException. I think in general, it's better to match behavior of non-federated client. If a path does not exist, just throw back FileNotFoundException.

I just verified symlinks. When target deleted,  ls on symlink does not throw FNFE.  Instead it is converted to file link. I tested dir->dir link. 
It seems this behavior is correct when compared with other fs. I tested on my MAC.
Should this be fixed in federated clusters is necessary ? Could you please validate this?
If we attempt to open that non existent link file, then we can throw out exception. But ls seems to simply pass.
```
Work % mkdir linkTarget 
Work % ln -s linkTarget linkSrc
Work % ls -l
lrwxr-xr-x   1 umagangumalla  xxxx     10 May 26 11:08 linkSrc -> linkTarget
Work % rm -rf linkTarget 
Work % ls -l            
lrwxr-xr-x   1 umagangumalla  xxxx     10 May 26 11:08 linkSrc -> linkTarget
Work % cd linkSrc
cd: no such file or directory: linkSrc
```